openassessment/xblock/static/js/src/lms/oa_response.js

```diff
@@ -487,47 +488,35 @@ export class ResponseView {
 
       const view = this;
       const { baseView } = this;
-      // eslint-disable-next-line new-cap
-      let fileDefer = $.Deferred();
+      let submitDefer = $.Deferred();
 
+      // Block submit if a learner has files staged for upload
       if (view.hasPendingUploadFiles()) {
-        if (!view.hasAllUploadFiles()) {
-          return;
-        }
-        const msg = gettext('Do you want to upload your file before submitting?');
-        if (window.confirm(msg)) {
-          fileDefer = view.uploadFiles();
-          if (fileDefer === false) {
-            return;
-          }
-        }
-      } else {
-        fileDefer.resolve();
+        return;
       }
 
-      fileDefer
-        .pipe(() => view.confirmSubmission()
-        // On confirmation, send the submission to the server
-        // The callback returns a promise so we can attach
-        // additional callbacks after the confirmation.
-        // NOTE: in JQuery >=1.8, `pipe()` is deprecated in favor of `then()`,
-        // but we're using JQuery 1.7 in the LMS, so for now we're stuck with `pipe()`.
-          .pipe(() => {
-            const submission = view.response();
-            baseView.toggleActionError('response', null);
-
-            // Send the submission to the server, returning the promise.
-            return view.server.submit(submission);
-          }))
-
-      // If the submission was submitted successfully, move to the next step
+      view.confirmSubmission()
+        // Proceed with submission
+        .done(() => {
+          const submission = view.response();
+          baseView.toggleActionError('response', null);
+
+          // Send the submission to the server, async
+          submitDefer = view.server.submit(submission);
```

Comment:
I've been looking at this and line 512 and I can't prove that there's something wrong with this, but I can't shake the feeling.

Reply:
so submitDefer is now the promise returned by server.submit(), but hasn't line 212 already executed by this point? I might just be confusing myself. Any chance you'd put that `.done().fail()` just directly on the `server.submit()` call?  we don't really need the defer variable anymore, right?

Reply:
I wrote it this way because I find the "flatter" layout more readable but yes, we could nest the `done` and `fail` calls. 

IMHO, it would be even better if we changed the `confirmSubmission` to not use a `$.Deferred()` since it's just wrapping a `window.confirm` and doesn't really _need_ to be async. That way we could have just the one layer of promises on the `server.submit()` call.